File: Sources/UI/Components/EdgeMark.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import SwiftUI

/// The intensity ladder's ONE mapping to colour, beside the mark that wears it. Three
/// surfaces draw the rung (Today's card, import preview, share sheet), so the judgement
/// lives in `PlanMath.IntensityBand` and the paint here, stated once.
extension PlanMath.IntensityBand {
    var tint: Color {
        switch self {
        case .unknown: Accent.bleu
        case .light: Accent.moss
        case .moderate: StatusTint.armed
        case .nearMax: Accent.alarm
        }
    }
}

/// The routine's signature grip on the bleu edge — the card's mark.
///
/// It began as the app icon verbatim, and every routine wore the same badge. Now it draws
/// the routine's SIGNATURE grip in `FingerGlyph`'s vocabulary, so cards differ exactly
/// when routines do — still one mark per card, never the per-set inventory. The SUMMARY
/// picks the signature (`RoutineSummary.signatureFingers`); the mark just draws it. The
/// rung is the one place bleu appears at rest: the icon's one accent, the edge the force
/// passes through.
struct EdgeMark: View {
    var fingers: FingerSet = .four
    /// The rung doubles as the INTENSITY signal (Nuri, 2026-08-17) — see
    /// `RoutineCard.rungTint`. The caller maps the band; the mark wears it.
    var rungTint: Color = Accent.bleu

    /// Scales with the title it sits beside; the declared size is the design size.
    @ScaledMetric(relativeTo: .title2) private var barWidth: CGFloat = 5.5

    private var gap: CGFloat { barWidth * 0.55 }
    /// **A deliberate departure from `HandGeometry.barAspect`**: at this small size the
    /// house ratio rounds bars into DOTS, one row above the session dots — the "circles all
    /// the way down" collision. Elongation keeps a bar a bar as it shrinks.
    private static let smallMarkAspect: CGFloat = 2.2
    private var tallestBar: CGFloat { barWidth * Self.smallMarkAspect }
    private var handWidth: CGFloat { barWidth * 4 + gap * 3 }

    var body: some View {
        VStack(alignment: .leading, spacing: barWidth * 0.32) {
            // Bottom-aligned at a common fingertip line: the tips are what is ON the edge.
            HStack(alignment: .bottom, spacing: gap) {
                ForEach(Array(fingers.occupied.indices), id: \.self) { index in
                    bar(on: fingers.occupied[index], index: index)
                }
            }
            if fingers.hasThumb {
                // The thumb lies down under the index side, as in `FingerPips`/`FingerGlyph`.
                RoundedRectangle(cornerRadius: barWidth * 0.31, style: .continuous)
                    .fill(Accent.graphite)
                    .frame(width: barWidth * 2 + gap, height: barWidth * 0.62)
            }
            RoundedRectangle(cornerRadius: barWidth * 0.36, style: .continuous)
                .fill(rungTint)
                .frame(width: handWidth, height: barWidth * 0.72)
        }
        // Decoration to a screen reader; the routine's NAME is the identity.
        .accessibilityHidden(true)
    }

    @ViewBuilder
    private func bar(on: Bool, index: Int) -> some View {
        let shape = RoundedRectangle(cornerRadius: barWidth / 2, style: .continuous)
        Group {
            if on {
                shape.fill(Accent.graphite)
            } else {
                // Stroked, full tertiary — the measured reason is in `FingerGlyph`.
                shape.strokeBorder(Ink.tertiary, lineWidth: 1)
            }
        }
        .frame(width: barWidth, height: tallestBar * HandGeometry.lengthFactor[index])
    }
}

#Preview {
    VStack(alignment: .leading, spacing: 20) {
        ForEach([FingerSet.four, .frontThree, .frontTwo, .backTwo], id: \.rawValue) { set in
            HStack(spacing: 16) {
                EdgeMark(fingers: set)
                Text("Daily no-hangs")
                    .font(.system(.title2, weight: .semibold))
            }
        }
    }
    .padding()
    .background { AppBackground() }
}

File: android/app/src/main/kotlin/run/nuri/getagrip/ui/components/EdgeMark.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import run.nuri.getagrip.engine.FingerSet
import run.nuri.getagrip.engine.PlanMath
import run.nuri.getagrip.ui.theme.GetAGripTheme
import run.nuri.getagrip.ui.theme.GripPalette
import run.nuri.getagrip.ui.theme.LocalGripPalette

/// The intensity ladder's ONE mapping to colour, kept beside the mark that wears it.
///
/// Several surfaces draw the rung, and a routine changing colour between two of them would
/// be the mark contradicting itself — so the JUDGEMENT lives in `PlanMath.IntensityBand`
/// (where a test pins the boundaries) and the PAINT lives here, stated once.
///
/// TRANSLATION NOTE: iOS writes this as an extension on the enum, which can reach its
/// colour constants directly. Kotlin's palette is a composition local, so the mapping takes
/// one instead — same single statement, one argument longer.
fun PlanMath.IntensityBand.tint(palette: GripPalette): Color = when (this) {
    PlanMath.IntensityBand.unknown -> palette.bleu
    PlanMath.IntensityBand.light -> palette.moss
    PlanMath.IntensityBand.moderate -> palette.armed
    PlanMath.IntensityBand.nearMax -> palette.alarm
}

/// The routine's signature grip on a tinted edge — the card's mark.
///
/// It began as the app icon verbatim (all four bars, always) and Nuri's first reaction named
/// the flaw: every routine wore the same badge. So the mark draws the routine's OWN
/// signature grip — filled where a finger is on the edge, hairline where it is not, the
/// thumb as its horizontal bar — in exactly `FingerGlyph`'s vocabulary, so cards differ
/// precisely when routines differ.
///
/// The rung is the one place an accent appears at rest — the icon's own rule: one accent
/// element, the edge the force passes through. It doubles as the routine's INTENSITY signal
/// (Nuri, 2026-08-17), and colour is reinforcement rather than the sole carrier: the plan
/// row still states the percentage in words.
@Composable
fun EdgeMark(
    fingers: FingerSet = FingerSet.four,
    modifier: Modifier = Modifier,
    barWidth: Dp = 5.5.dp,
    rungTint: Color = LocalGripPalette.current.bleu,
) {
    val palette = LocalGripPalette.current
    val gap = barWidth * 0.55f
    // 2.2×, deliberately more elongated than `FingerGlyph`'s 1.75: at this mark's small
    // size the glyph's ratio rounds the bars into DOTS — and four filled dots one row above
    // the session dots is the "circles all the way down" collision the shape vocabulary
    // exists to prevent. Elongation is what keeps a bar a bar when the drawing shrinks.
    val tallest = barWidth * 2.2f
    val handWidth = barWidth * 4 + gap * 3

    Column(
        modifier.clearAndSetSemantics {},
        verticalArrangement = Arrangement.spacedBy(barWidth * 0.32f),
    ) {
        // Bottom-aligned: the varying lengths meet at a common fingertip line, because the
        // tips are what is ON the edge below.
        Row(
            horizontalArrangement = Arrangement.spacedBy(gap),
            verticalAlignment = Alignment.Bottom,
        ) {
            fingers.occupied.forEachIndexed { index, on ->
                val shape = RoundedCornerShape(barWidth / 2)
                Box(
                    Modifier
                        .width(barWidth)
                        .height(tallest * LENGTH_FACTOR[index])
                        .then(
                            if (on) {
                                Modifier.background(palette.graphite, shape)
                            } else {
                                // Stroked, not tinted lighter — fill-vs-outline survives
                                // greyscale, and full tertiary ink because a 1 dp hairline
                                // is nearly all antialiased edge.
                                Modifier.border(1.dp, palette.inkTertiary, shape)
                            },
                        ),
                )
            }
        }
        if (fingers.hasThumb) {
            // The thumb lies down under the index side, exactly as `FingerPips` and
            // `FingerGlyph` draw it — bars are digits; this one is horizontal.
            Box(
                Modifier
                    .width(barWidth * 2 + gap)
                    .height(barWidth * 0.62f)
                    .background(palette.graphite, RoundedCornerShape(barWidth * 0.31f)),
            )
        }
        Box(
            Modifier
                .width(handWidth)
                .height(barWidth * 0.72f)
                .background(rungTint, RoundedCornerShape(barWidth * 0.36f)),
        )
    }
}

/// A HAND's proportions — `FingerGlyph`'s array, verbatim.
private val LENGTH_FACTOR = listOf(0.86f, 1.0f, 0.94f, 0.80f)

@Preview(name = "EdgeMark", showBackground = true, widthDp = 320)
@Composable
private fun EdgeMarkPreview() {
    GetAGripTheme {
        val palette = LocalGripPalette.current
        Row(
            Modifier.padding(20.dp),
            horizontalArrangement = Arrangement.spacedBy(24.dp),
            verticalAlignment = Alignment.Bottom,
        ) {
            EdgeMark(FingerSet.four, barWidth = 10.dp)
            EdgeMark(
                FingerSet.frontTwo,
                barWidth = 10.dp,
                rungTint = PlanMath.IntensityBand.light.tint(palette),
            )
            EdgeMark(
                FingerSet.frontTwo.union(FingerSet.thumb),
                barWidth = 10.dp,
                rungTint = PlanMath.IntensityBand.nearMax.tint(palette),
            )
        }
    }
}
